sycl/source/detail/scheduler/commands.cpp

```diff
@@ -1196,12 +1196,8 @@ AllocaCommandBase *ExecCGCommand::getAllocaForReq(Requirement *Req) {
   throw runtime_error("Alloca for command not found", PI_INVALID_OPERATION);
 }
 
-void ExecCGCommand::flushStreams() {
-  assert(MCommandGroup->getType() == CG::KERNEL && "Expected kernel");
-  for (auto StreamImplPtr :
-       ((CGExecKernel *)MCommandGroup.get())->getStreams()) {
-    StreamImplPtr->flush();
-  }
+vector_class<StreamImplPtr> ExecCGCommand::getStreams() const {
+  return ((CGExecKernel *)MCommandGroup.get())->getStreams();
```

Comment:
According to description of the patch
> `((ExecCGCommand *)NewCmd)->flushStreams();` looks not thread safe.

Why is it used here, then?

Reply:
Yes, let me explain. 
If you use sycl stream class to print data from kernel, then in the `Scheduler::addCG()`, the `ExecCGCommand::flush()` function will be executed.
Here we had two problems:  1) using NewCmd that can be invalid
2) We can't just use `ExecCGCommand::flushStreams()` under locked `MGraphLock` mutex.
Since `Scheduler::addHostAccessor()` will be called inside `ExecCGCommand::flushStreams()`,
and the `Scheduler::addHostAccessor()` also locks the `MGraphLock` and that can lead to deadlock. 
So, to void deadlock, I changed  `ExecCGCommand::flushStreams()` to  `ExecCGCommand::getStreams()`
and first, we get the streams when we under locked `MGraphLock` mutex, and then without this mutex, we flush all streams.